specify/specify7
# Avoid query duplicates by grouping by only display fields

Fixes #1450

This PR is based on the [issue-1450-2](https://github.com/specify/specify7/tree/issue-1450-2) branch, but also proposes the [issue-1450-3](https://github.com/specify/specify7/tree/issue-1450-3) branch as another potential solution.

Here is an example of the query result from the query described in the #1450 issue
```
[61087,"101010104","Balistidae","Abalistes",null,"K & R 6","04/21/1964","<Aggregator not defined.>","North America, United States, Missouri, Platte","Missouri River, beneath Fairfax Bridge","39.1580429077","-94.6241989136","<Aggregator not defined.>",null,null,"Point","0E-10",2846]
[61087,"101010104","Balistidae","Abalistes",null,"K & R 6","04/21/1964","<Aggregator not defined.>","North America, United States, Missouri, Platte","Missouri River, beneath Fairfax Bridge","39.1580429077","-94.6241989136","<Aggregator not defined.>",null,null,"Point","0E-10",2846]
```
After the changes made in this branch, just one of the duplicate rows is returned
```
[61087,"101010104","Balistidae","Abalistes",null,"K & R 6","04/21/1964","<Aggregator not defined.>","North America, United States, Missouri, Platte","Missouri River, beneath Fairfax Bridge","39.1580429077","-94.6241989136","<Aggregator not defined.>",null,null,"Point","0E-10",2846]
```

The motive of this solution is to avoid the user seeing multiple identical rows next to each other in the query results.  In order to avoid duplicate records like in #1450, this solution takes all of the display fields checked by the user and adds them to a GROUP BY statement, thus avoiding duplicates.

This solution's idea came from https://github.com/specify/specify7/issues/1464#issuecomment-1286985113
> When running a distinct query, can we mimic a distinct query using "GROUP BY" instead of "SELECT DISTINCT" and include the id in the query as a `CONCAT(id, ", ")` (which would be parsed on the front-end into separate records)?

I think this solution is a good idea, and implementing a front-end solution to handle a `GROUP_CONCAT(id, ", ")` field in the record set could be nice.  See branch [issue-1450-3](https://github.com/specify/specify7/tree/issue-1450-3) and commit [group_concat record id field for distinct queries](https://github.com/specify/specify7/commit/92b1a129d513e9f9f92bfa7eaeffa798af303358) where I coded the back-end part for this solution, but still needs front-end code to handle multiple record IDs.

The difference from this branch, is that the issue-1450-3 branch moves the record ID from the group by statement to the GROUP_CONCAT statement.  When showing the results, each row will need to provide links to all the records. After talking with Andy, he thought that checking the distinct checkbox would work if it provided links to each record.  The current design could be modified, but we need to be mindful of any other side effects in changing the original design intentions https://github.com/specify/specify7/blob/67d73f4663cfef67356e95487da61cae1ae43db9/specifyweb/stored_queries/execution.py#L577

Please post your opinion on which solution we should go with for now, branch `issue-1450-2` or branch `issue-1450-3`?
I'm thinking we should further develop the `issue-1450-3` branch.


### Checklist

      and self-explanatory (or properly documented)

### Testing instructions

Use the query builder to make queries like those in #1450 and make sure that there are no duplicate records in the results.
